feat: add process docs and scripts for yarn.lock only backports (#2713)
redhat-developer/rhdh-plugin-export-overlays: user-guide/06-patch-management.md

Step 3 only covers what step 1–2 did not — lockfile bumps without another upstream release.
Re-run `generate` with the full CVE list after step 2 so `repo_ref` in the manifest matches
the updated baseline.

```
workspaces/[ws]/patches/
├── 0-cve-yarn-lock.patch    # applied first (max one)
├── cve-backports.yaml       # auto-generated
└── 1-fix-something.patch    # optional code/build patches
```

Tooling: `scripts/yarnlock-backport/` (`npm install` once). Requires Node.js
(see `versions.json`), `git`, `yarn`, `patch`, `diff`, `npm`. Use after step 2 above
when transitive bumps are still required at the current `repo-ref`.

| Path / remote | Requirement |
|---------------|-------------|
| `--overlay-workspace` | Absolute path; fork needs `upstream` on rhdh-plugin-export-overlays (prepare syncs `release-{version}`) |
| `--plugins-repo` | Absolute path to a local clone of **`source.json:repo`** (e.g. rhdh-plugins, [community-plugins](https://github.com/backstage/community-plugins)); git remote is resolved from that URL |

Use a [git worktree](https://git-scm.com/docs/git-worktree) on the overlay release branch when patch files are not on your current branch.

> ```bash

Reviewer: ```suggestion
e.g.
```bash
# Overlay release branch (for patches/manifest on release-1.10)
git worktree add ../rhdh-overlays-release-1.10 release-1.10
# Optional: dedicated tree for script development (keeps main checkout clean)
git worktree add ../rhdh-overlays-scripts lockfile-patch-scripts
```
It wasn't obvious to me how to set up the git worktrees.  Can we add the above lines for clarity?

cd scripts/yarnlock-backport && npm install   # local tool — not published to npm

> export OVERLAY_WORKSPACE=<absolute-path>/workspaces/orchestrator

Reviewer: ```suggestion
export OVERLAY_WORKSPACE=<absolute-git-worktree-path>/workspaces/orchestrator
```

export PLUGINS_REPO=<absolute-path>   # clone matching source.json:repo

npx yarnlock-backport prepare --release 1.10 \
  --overlay-workspace "$OVERLAY_WORKSPACE" --plugins-repo "$PLUGINS_REPO"

> # Bump deps in $PLUGINS_REPO/workspaces/<ws>/ (yarn up, triager, etc.)

Reviewer: Maybe we should specify this is a manual step with TBD instructions.  We still need to flesh out the instructions for plugin owners

> # Bump deps in $PLUGINS_REPO/workspaces/<ws>/ (yarn up, triager, etc.)

Reviewer: ```suggestion
# Manual step: Update dependencies in $PLUGINS_REPO/workspaces/<ws>/ (Instructions TBD)
```

npx yarnlock-backport generate --release 1.10 \
  --overlay-workspace "$OVERLAY_WORKSPACE" --plugins-repo "$PLUGINS_REPO" \
  --cve 'CVE-2026-44487,CVE-2026-41674/@xmldom/xmldom'
```

| Command | What it does |
|---------|--------------|
| **prepare** | Sync overlay release branch → checkout `repo-ref` in source repo → `yarn install` → apply existing `0-cve-yarn-lock.patch` |
| **generate** | `yarn install` → diff `repo-ref` baseline vs local `yarn.lock` → write patch + merge manifest (fetches MITRE) |

**Flags:** `--release`, `--overlay-workspace`, `--plugins-repo` (required). `prepare`: `--skip-patch`, `--force`, `--verbose`, `--dry-run`. `generate`: `--cve` (required), `--dry-run`.

> **`--cve` syntax** (same as [rhdh-security triager](https://github.com/redhat-developer/rhdh-security/blob/main/triage/triager.py)): comma-separated CVE ids. Append `/npm-package` when MITRE names the product differently from the npm package (e.g. `CVE-2026-41242/protobufjs`). Multiple npm aliases: `CVE-2026-41674/@xmldom/xmldom,xmldom`.

Reviewer: You are referencing a script from our GL repo.  Are you planning to expose publicly?  If so, perhaps we should make it available in the https://github.com/redhat-developer/rhdh-skill/tree/main/skills/rhdh-jira repo

Author: Removed

**Debugging:** `yarn install` streams output by default. Use `--verbose` on prepare to also stream git fetch/checkout. Other subprocess failures print captured stderr plus exit code and `cwd`.

> **Re-roll:** If prepare fails at patch apply, discard the stale patch, restore bumps from `cve-backports.yaml` (`yarn up`), apply any new fixes, then re-run **generate** with the **full** `--cve` list.

Reviewer: > restore bumps from `cve-backports.yaml` (`yarn up`)

We should not be so specific here.  "restore bumps" in the Z stream will likely mean using `yarn-lockfile-surgeon` to update to the previously shipped versions.  I would rather we have a TBD placeholder for now while we figure out the details.

> **Re-roll:** If prepare fails at patch apply, discard the stale patch, restore bumps from `cve-backports.yaml` (`yarn up`), apply any new fixes, then re-run **generate** with the **full** `--cve` list.

Reviewer: ```suggestion
**Re-roll:** If prepare fails at patch apply, discard the stale patch, restore bumps from `cve-backports.yaml` (Instructions TBD), apply any new fixes, then re-run **generate** with the **full** `--cve` list.
```

**Verify (future):** Planned `yarnlock-backport verify` against OCI images after `/publish`.

### Manifest (`cve-backports.yaml`)

Rows merge CVEs sharing the same `package` + `patch_version`. `repo_ref` is `source.json:repo-ref` at generate time (patch baseline). `patch_version` is comma-separated when multiple major lines exist.

```yaml
patch_file: 0-cve-yarn-lock.patch
repo_ref: eb6cce6110fc8bd532e717e9d995764481b36f1b
backports:
  - cve_ids: [CVE-2026-44486, CVE-2026-44487]
    package: axios
    patch_version: 1.18.1
    notes: Upstream PR 3537; shipped in backend OCI
  - cve_ids: [CVE-2026-45736, CVE-2026-48779]
    package: ws
    patch_version: 8.18.0, 8.21.0
    notes: |-
      [auto] ws@8.18.0 is still in CVE affected range; verify dev-only
      └─┬ @internal/orchestrator@1.0.0
        └── … → ws@8.18.0
```

> **`[auto]` notes:** On generate, each `patch_version` is checked against MITRE affected ranges ([`branch-check.py`](https://github.com/redhat-developer/rhdh-security/blob/main/triage/branch-check.py) logic). Versions still in range get a warning plus an `npm ls` spine (usually dev-only). Manual notes above `[auto]` are preserved; the `[auto]` block is refreshed each run. Confirm with `branch-check.py` before merging.

Reviewer: branch-check.py repo does not exist

Author: Removed reference 👍

| Event | Action |
|-------|--------|
| First backport | generate → commit patch + manifest |

> | Add CVEs / re-roll | `yarn up` fixes → generate with full `--cve` list |

Reviewer: ```suggestion
| Add CVEs / re-roll | update dependencies (Instructions TBD) → generate with full `--cve` list |
```

export OVERLAY_WORKSPACE=<absolute-path>/workspaces/orchestrator
export PLUGINS_REPO=<absolute-path>
```

Both paths must be absolute. When testing tooling on a feature branch while the
overlay content lives on a release branch, point `--overlay-workspace` at a
[git worktree](https://git-scm.com/docs/git-worktree) checked out at
`release-{version}` (prepare step 0 syncs the overlay repo root to that branch).

**Step 1 — Prepare plugins workspace**

Reset the plugins workspace to the pinned `repo-ref` and leave `yarn.lock` in a
known-good state — either with the existing CVE patch applied, or ready for new
fixes in Step 2.

```bash
yarnlock-backport prepare \
  --release 1.10 \
  --overlay-workspace "$OVERLAY_WORKSPACE" \
  --plugins-repo "$PLUGINS_REPO"
```

Sub-steps (all in prepare):

| Step | Action |
|------|--------|
| **0** | Fetch and fast-forward overlay repo branch `release-{release}` from `upstream` (or `origin` on a direct clone) |
| **1** | Checkout plugins repo at `repo-ref` from `source.json` |
| **1a** | `yarn install` |
| **1b** | Apply `0-cve-yarn-lock.patch` if present |

| Option | Required | Description |
|--------|----------|-------------|

> | `--release` | yes | RHDH release line (e.g. `1.10` → branch `release-1.10` on rhdh-plugin-export-overlays) |

Reviewer 2: non-blocking comment: if 1.10 means release-1.10, why dont the user specify direclty --release release-1.10

  --plugins-repo "$PLUGINS_REPO" \
  --cve 'CVE-2026-44487,CVE-2026-44494'
```

| Option | Required | Description |
|--------|----------|-------------|
| `--release` | yes | RHDH release line (e.g. `1.10` → branch `release-1.10` on rhdh-plugin-export-overlays) |
| `--overlay-workspace` | yes | Absolute path to the overlays workspace; reads `source.json` and `patches/` |
| `--plugins-repo` | yes | Absolute path to the local rhdh-plugins repository root |
| `--cve` | yes | Comma-separated CVE ids to fetch from MITRE and record in the manifest; optional `/package` override per id (e.g. `CVE-2026-1234/axios`) |
| `--dry-run` | no | Validate inputs and print actions without writing the patch or manifest |

**Step 4 — Verify OCI artifacts (future)**

Planned: `yarnlock-backport verify` to compare published OCI image
`node_modules` against the manifest after `/publish`.

### Manifest format (`cve-backports.yaml`)

Auto-generated by `yarnlock-backport generate`. CVEs that share the same
`package` and `patch_version` are **merged into one row** with a `cve_ids` list:

```yaml
patch_file: 0-cve-yarn-lock.patch

backports:
  - cve_ids:
      - CVE-2026-44486

>       - CVE-2026-44487

Reviewer 2: non-blocking: in the example we should set at least one cve with override
